README.md

```diff
@@ -87,6 +87,50 @@ The PMEM-CSI driver supports running in different modes, which can be controlled
+down. Data is retained until the volume gets deleted.
+* Ephemeral Volumes
+An application is assigned to a node and when it starts to run, a new volume is created for it on that node. When the application stops, the volume is deleted. This [diagram](/docs/images/sequence/pmem-csi-ephemeral-sequence-diagram.png) illustrates how an ephemeral volume gets provisioned in Kubernetes using PMEM-CSI driver.
+* Cache Volumes
+Volumes are created on a certain set of nodes, each with its own local data. Applications are started on those nodes and then get to use the volume on their node. Data persists across application restarts. This
+is useful when the data is only cached information that can be discarded and reconstructed at any time *and* the application can reuse existing local data when restarting. This [diagram](/docs/images/sequence/pmem-csi-cache-sequence-diagram.png) illustrates how a cache volume gets provisioned in Kubernetes using PMEM-CSI driver.
+
+Volume | Kubernetes | PMEM-CSI | Limitations
+--- | --- | --- | ---
+Persistent | supported | supported | topology aware scheduling
+Ephemeral | [in design](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190122-csi-inline-volumes.md#proposal) | supported | topology aware scheduling, resource constraints
+Cache | supported | supported | topology aware scheduling
+
+[Topology aware scheduling](https://github.com/kubernetes/enhancements/issues/490)
+ensures that an application runs on a node where the volume was created. This is obviously needed for persistent volumes, but also for
+delayed persistent volumes because otherwise a restarted application
+might end up on the wrong node. Usage of cache volumes is more flexible,
+but an application still can only run on one of the nodes where the
+a cache volume is available. For CSI-based drivers like PMEM-CSI,
+Kubernetes >= 1.13 is needed. On older Kubernetes releases, pods must
+be scheduled manually onto the right node(s).
+
+[Topology aware volume provisioning](https://kubernetes.io/blog/2018/10/11/topology-aware-volume-provisioning-in-kubernetes/)
+is needed for delayed persistent volumes.
+
+The upstream design for ephemeral volumes currently does not take
+[resource constraints](https://github.com/kubernetes/enhancements/pull/716#discussion_r250536632)
+into account. If an application gets scheduled onto a node and then
+creating the ephemeral volume on that node fails, the application on
+the node cannot start until resources become available. The same
+applies to how this is currently implemented in PMEM-CSI: a fake
+volume gets created initially, without actually allocating space. Then
+when an application starts and it is time to mount that "existing"
+volume, an attempt is made to allocate the space required for the
+volume locally. Only if there is enough space can application startup
```

Comment:
According to https://github.com/intel/pmem-CSI/blob/046cadf8461d429203db204e46289140d295534f/docs/images/sequence/pmem-csi-ephemeral-sequence-diagram.png, ephemeral volumes are implemented such that they can only be used once: after unattaching it, the volume is gone. Trying to use it again for another app or restarting the app will fail (volume is not going to be re-created because that happens during `CreateVolume`).

What I described in https://github.com/intel/pmem-CSI/pull/160/files#r254731929 is different. It would allow the PVC to be created once and then be used multiple times.

This is a choice between a rock and a hard place: the current implementation requires more work from the user and probably can't work when pods get restarted by a controller (because the controller doesn't know that it needs to re-created a PVC), while reusing the same PVC has the drawback that it ignores local volume availability during pod scheduling (as documented above).

@avalluri is it correct that with the current implementation pods will always land on a node which had enough space? I'm still a bit fuzzy about that aspect of pod scheduling when WaitForFirstConsumer is active, and I haven't checked whether your implementation actually looks for a node with free space. We need to check what happens when Kubernetes picks one node for the pod and then the volume gets created on another node because the one picked by Kubernetes didn't have enough space.

I'm open for debate which one we should implement. @okartau, any opinion?

Reply:
> According to https://github.com/intel/pmem-CSI/blob/046cadf8461d429203db204e46289140d295534f/docs/images/sequence/pmem-csi-ephemeral-sequence-diagram.png, ephemeral volumes are implemented such that they can only be used once: after unattaching it, the volume is gone. Trying to use it again for another app or restarting the app will fail (volume is not going to be re-created because that happens during `CreateVolume`).
> 
> What I described in https://github.com/intel/pmem-CSI/pull/160/files#r254731929 is different. It would allow the PVC to be created once and then be used multiple times.
> 
Expecting re-attachable volumes in case of "ephemeral" is not a right ask in my opinion, The current implementation supports application restarts by kubelet, as the VolumeAttachement object wont get deleted in this case.

But if we really wants to implement reattachable ephemeral volumes, what we can do best is to clean data on NodeUnpulbish/Unstage instead of deleting the device, in that case the space is not freed even the application has stopped, till the PVC gets deleted.

> @avalluri is it correct that with the current implementation pods will always land on a node which had enough space? I'm still a bit fuzzy about that aspect of pod scheduling when WaitForFirstConsumer is active

The current implementation always  try to create volume(s) on nodes which have free space in the order of requested *AccessibilityRequirement*. We have no information if that node is really picked by kubernetes(delayed binding) or not.

Reply:
"The current implementation supports application restarts by kubelet, as the VolumeAttachement object wont get deleted in this case." - I doubt that app restarts work as well you seem to think it does.

Consider the following case:
- An app with an ephemeral volume starts on one node.
- The node is marked as unschedulable in preparation for taking it down.
- The app is restarted.
- The expectation is that it starts up on another node.

Do you agree that this doesn't work with what you have implemented?

Let me phrase it differently: the definition of "ephemeral volume" is that an app gets a new volume each time it starts and (let me add that for clarity) regardless where it starts.

Reply:
It seems that you understood "ephemeral" more in the sense of "one-time use", right?

Re-reading my original description ("An application is assigned to a node and when it starts to run, a new volume is created for it on that node. When the application stops, the volume is deleted.") I see that two words are missing. It should have said "*Each time* an application...". Sorry if that wasn't clear.

Reply:
Here's another use-case for ephemeral volumes: running multiple instances of the same app concurrently, each with its own local PMEM storage.

The way to specify that for Kubernetes is with a ReplicaSet. An inline volume would be the natural way to specify the volume, but it's not currently supported, so we have to use a PVC. The end result then looks something like this:
```
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: frontend
  labels:
    app: guestbook
    tier: frontend
spec:
  # modify replicas according to your case
  replicas: 3
  selector:
    matchLabels:
      tier: frontend
  template:
    metadata:
      labels:
        tier: frontend
    spec:
      containers:
      - name: php-redis
        image: gcr.io/google_samples/gb-frontend:v3
      volumeMounts:
      - mountPath: "/scratch"
        name: my-csi-volume
      volumes:
      - name: my-csi-volume
        persistentVolumeClaim:
          claimName: php-redis-volume
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: php-redis-volume
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 8Gi
storageClassName: pmem-csi-sc-ephemeral
```

After considering this example, I need to revise my earlier proposal from https://github.com/intel/pmem-CSI/pull/160/files#r254731929: we don't want two pods running on the same node to share data. That would be unexpected. That means that volume creation has to be done in `NodePublishVolume` and deletion in `NodeUnpublishVolume`.

Reply:
> It seems that you understood "ephemeral" more in the sense of "one-time use", right?
> 
If i am not wrong, the ephemeral volume is for *one-time-use", creation/deletion of volume(PV) is not in driver scope, its COs job to request at app startup and shutdown. We know that currently its not supported in Kubernetes, So what i did is "in case of ephemeral delete the pmem volume on *ControllerUnpublish*, instead of waiting for *DeleteVolume* request.

Reply:
> "The current implementation supports application restarts by kubelet, as the VolumeAttachement object wont get deleted in this case." - I doubt that app restarts work as well you seem to think it does.
> 
> Consider the following case:
> 
> * An app with an ephemeral volume starts on one node.
> * The node is marked as unschedulable in preparation for taking it down.
> * The app is restarted.
> * The expectation is that it starts up on another node.
> 
> Do you agree that this doesn't work with what you have implemented?
In my humble opinion this is not in scope of our driver, CO/user has to request new PV on app startup.

Reply:
> After considering this example, I need to revise my earlier proposal from https://github.com/intel/pmem-CSI/pull/160/files#r254731929: we don't want two pods running on the same node to share data.

@pohly This case has been already considered by current implementation by rejecting multiple *NodePublishVolume* requests for ephemeral volumes. Please check with the implementation if its needed any changes.

Reply:
> In my humble opinion this is not in scope of our driver, CO/user has to request new PV on app startup.
> ...
> It rejects multiple NodePublishVolume requests for ephemeral volumes.

This is the crux of the matter: this implementation of a one-time volume isn't very useful for users. The goal was to implement something that is as close as possible to an inline ephemeral volume, because "give me local scratch space for my app" is used by apps today.

What you are saying is that users need to rewrite their app deployment logic and can't use controllers like `ReplicaSet`. Why not change the implementation so that this all works? Is there some advantage of this implementation over what I was suggesting?

Reply:
> What you are saying is that users need to rewrite their app deployment logic and can't use controllers like `ReplicaSet`. 

"cache" volume are best suits for `ReplicaSet/Deployment` . Pre-create the cached PVC with no of replicas and ask to use that PVC by ReplicaSet.

and "ephemeral" volumes are supposed to use by `StatefulSet`, because StatefulSets  supports `volumeClaimTemplate` where each volume tied to a PVC, the volumes are not shared.